File: modules/web/app/model/task/Condition.java
package model.task;

import javax.persistence.Column;
import javax.persistence.EmbeddedId;
import javax.persistence.Entity;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;

import io.ebean.Finder;
import io.ebean.Model;

@Entity
@Table(name = "conditions")
public class Condition extends Model {

  public static final Finder<JsConditionKey, Condition> finder = new Finder<>(Condition.class);

  @EmbeddedId
  public JsConditionKey key;

  @ManyToOne
  @JoinColumn(name = "task_id", referencedColumnName = "task_id", insertable = false, updatable = false)
  @JoinColumn(name = "exercise_id", referencedColumnName = "exercise_id", insertable = false, updatable = false)
  @JsonBackReference
  public JsWebTask task;

  @Column
  public String xpathQuery;

  @Column
  public boolean isPrecondition;

  @Column
  public String awaitedValue;

  public Condition(JsConditionKey theKey) {
    key = theKey;
  }

  @JsonIgnore
  public String getDescription() {
    return "Element mit XPath \"" + xpathQuery + "\" sollte den Inhalt \"" + awaitedValue + "\" haben";
  }

  @JsonIgnore
  public boolean isPostcondition() {
    return !isPrecondition;
  }

  @JsonIgnore
  public boolean isPrecondition() {
    return isPrecondition;
  }

}

File: app/model/web/Condition.scala
package model.web

import com.fasterxml.jackson.annotation.JsonIgnore
import play.api.db.slick.HasDatabaseConfigProvider
import slick.jdbc.JdbcProfile

case class Condition(id: Int, taskId: Int, exerciseId: Int, xpathQuery: String, isPrecondition: Boolean, awaitedValue: String) {

  @JsonIgnore
  def description =
    s"""Element mit XPath "$xpathQuery" sollte den Inhalt "$awaitedValue" haben"""

}

private[model] trait Conditions extends WebTasks {
  self: HasDatabaseConfigProvider[JdbcProfile] =>

  import profile.api._

  class ConditionsTable(tag: Tag) extends Table[Condition](tag, "CONDITIONS") {

    def conditionId = column[Int]("CONDITION_ID")

    def taskId = column[Int]("TASK_ID")

    def exerciseId = column[Int]("EXERCISE_ID")

    def xpathQuery = column[String]("XPATH_QUERY")

    def isPrecondition = column[Boolean]("IS_PRECONDITION")

    def awaitedValue = column[String]("AWAITED_VALUE")


    def pk = primaryKey("PK", (conditionId, taskId, exerciseId))

    def taskFk = foreignKey("TASK_FK", (taskId, exerciseId), jsTasks)(t => (t.id, t.exerciseId))


    def * = (conditionId, taskId, exerciseId, xpathQuery, isPrecondition, awaitedValue) <> (Condition.tupled, Condition.unapply)

  }

  lazy val conditions = TableQuery[ConditionsTable]

}
